---
title: "Figure S2"
author: "Sergio Garcia-Moreno Alcantara (sga@stowers.org) "
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  pdf_document:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
    code_folding: "hide"
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-01444-003-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options: 
  chunk_output_type: inline
---

# Aim

The aim of this analysis is to define promoter types in Drosophila kc167 cells based on the presence of specific core promoter elements

# Enviroment setup

Set working directory and load required libraries and lab functions

```{r setup, warning=F, message=F, tidy=TRUE}

setwd("/n/projects/sga/analysis/SAGA/saga_publication/")
options(knitr.figure_dir = "plots/figure_s2/")

#Standard packages
library(GenomicAlignments); library(GenomicRanges); library(Biostrings); library(BSgenome.Dmelanogaster.UCSC.dm6);
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene); library(dplyr); library(reshape2); library(plyranges); library(CAGEr); 
library(magrittr); library(ggplot2); library(cowplot); library(ggseqlogo); library(gridExtra); library(ggpubr)


#Lab sources
#Lab sources
source("./shared_code/granges_common.r")
source("./shared_code/metapeak_common.r")
source("./shared_code/knitr_common.r")

```

# Analysis

## 1. Define promoter types 

```{r select promoter types, tidy=TRUE}

# Load TSS
tss <- get(load("./rdata/cage_kc167_tss_pLaw_2tpm.RData"))

## Define function to find promoter element (motifs) in each active tss
find_motif <- function(motif_name, fb_t_id,  mismatch=0) {
    
    motif_info <- subset(promoter_table, name == motif_name)
    motif <- DNAString(motif_info$motif)
    up_dis <- motif_info$window_start
    down_dis <- motif_info$window_end
    
    gene_tss <- tss[tss$fb_t_id %in% fb_t_id]
    
    if(up_dis >= 0 & down_dis >=0){
      tss_r <- resize(gene_tss, down_dis, "start") %>%
               resize(., down_dis - up_dis, "end")
    }
    if(up_dis < 0 & down_dis >=0){
      tss_r <- resize(gene_tss, down_dis, "start") %>%
               resize(., abs(up_dis)+down_dis, "end")
    }
    if(up_dis < 0 & down_dis <0){
      tss_r <- resize(gene_tss, abs(up_dis), "end") %>%
               resize(., abs(up_dis)-abs(down_dis), "start")
    }
    
    promoter_seq <- getSeq(Dmelanogaster, tss_r)
    names(promoter_seq) <- tss_r$fb_t_id
    
    count_df <- vcountPattern(motif, promoter_seq, fixed = FALSE, 
                              min.mismatch = 0, max.mismatch = mismatch) %>%
                data.frame(fb_t_id = fb_t_id, count =.)
    
    count_df$count <- ifelse(count_df$count >0, T, F)
    colnames(count_df)[2] <- motif_name
    count_df
}

## Provide promoter element (motif) search information (motif sequence composition and search window relative to the TSS)
promoter_table <- read.table("./promoter_elements_sga.txt", header=T)
motifs <- promoter_table$name

## Find motifs across TSSs allowing 0 and 1 mismatch

motif_list_1mm <- mclapply(as.character(motifs), function(x){
  motif <- find_motif(motif_name=x, tss$fb_t_id, 1)
  motif
  }, mc.cores =3)

motif_list_0mm <-   mclapply(as.character(motifs), function(x){
  motif <- find_motif(motif_name=x, tss$fb_t_id, 0)
  motif
  }, mc.cores =3)

motif_df_1mm <- reshape::merge_recurse(motif_list_1mm)
motif_df_0mm <- reshape::merge_recurse(motif_list_0mm)

save(motif_df_1mm, file="./rdata/motif_df_kc167_1mm.RData")
save(motif_df_0mm, file="./rdata/motif_df_kc167_0mm.RData")

tss_info <- as.data.frame(tss)[c(1:16)]

motif_info_df_0 <- merge(tss_info, motif_df_0mm)
motif_info_df_1 <- merge(tss_info, motif_df_1mm)

## Define promoter groups

tata_tss <- tss[tss$fb_t_id %in% subset(motif_df_1mm, TATA)$fb_t_id]
dpe_tss <- tss[tss$fb_t_id %in% subset(motif_df_1mm, !(TATA) & DPE_O | PB)$fb_t_id]
tct_tss <- tss[tss$fb_t_id %in% subset(motif_df_0mm, TCT)$fb_t_id]
hk_tss <- tss[tss$fb_t_id %in% subset(motif_df_0mm, !(TATA | TCT | MTE | DPE | DPE_K | DPE_O | PB | Inr) & (DRE | Motif1 | Motif6 | Motif7))$fb_t_id]

motif_list_kc167 <- list(tata=tata_tss, dpe=dpe_tss, tct=tct_tss, hk=hk_tss)
save(motif_list_kc167, file="./rdata/motif_list_kc167.RData")

```

## 2. Plot a DNA-sequence heatmap of the different promoters types

```{r sequence heatmap, warning=FALSE, fig.height=12, fig.width=3, tidy=TRUE}

## Define function
get_heatmap <- function(tss, window, direction, name){
  seq <- getSeq(Dmelanogaster, resize(tss, window, direction))
  seq_df <- as.character(seq) %>% lapply(., function(x)strsplit(x, "")) %>% 
    unlist(., recursive = F) %>% do.call(rbind,.) %>% as.data.frame()

  seq_df$id <- 1:nrow(seq_df)
  seq_df_m <- reshape2::melt(seq_df, id.vars  = "id")
  
  ATGC_plot <- ggplot(seq_df_m, aes(x = variable, y = id, fill = value)) + 
    geom_raster() +
    scale_fill_manual(values = c("indianred3", "#2C699B", "black", "#FDCC4E")) +
    xlab(paste(window, "bp around TSS")) + ylab("Genes") + 
    ggtitle(name) +
    theme_cowplot() +
    theme(axis.ticks.y = element_blank(),
          axis.text.y = element_blank(),
          axis.text.x = element_blank())
}

## Generate heatmaps
tata_hm <- get_heatmap(tata_tss, 101, "center", paste("TATA genes","(n =",length(tata_tss),")"))
dpe_hm <- get_heatmap(dpe_tss, 101, "center", paste("DPE genes","(n =",length(dpe_tss),")"))
tct_hm <- get_heatmap(tct_tss, 101, "center", paste("TCT genes","(n =",length(tct_tss),")"))
hk_hm <- get_heatmap(hk_tss, 101, "center", paste("HK genes","(n =",length(hk_tss),")"))

plot_grid(tata_hm, dpe_hm, tct_hm, hk_hm, ncol=1)

```

## 3. Plot a position weigth matrix (PWM) across the promoter types TSSs

```{r logo, warning=FALSE, fig.height=6, fig.width=16, tidy=TRUE}

## Define function
get_logo <- function(tss){
  cs2 = make_col_scheme(chars=c('A', 'T', 'C', 'G'), 
                        cols=c("indianred3", "#FDCC4E", "#2C699B", "black"))
  seq <- as.vector(getSeq(Dmelanogaster, resize(tss, 101, "center")))
  ggseqlogo(seq, col_scheme=cs2) +
  theme(axis.text.x = element_text(size = 6),
        axis.ticks.x = element_line())
}

## Plot logos
tata_logo <- get_logo(tata_tss)
dpe_logo <- get_logo(dpe_tss)
tct_logo <- get_logo(tct_tss)
hk_logo <- get_logo(hk_tss)

plot_grid(tata_logo, dpe_logo, tct_logo, hk_logo, ncol=1)

```

# Session Info

```{r}
sessionInfo()
```
